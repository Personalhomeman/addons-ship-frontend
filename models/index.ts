import { NextContext } from 'next';
import { Store } from 'redux';
import { Request } from 'express';

export type AppVersionProps = {
  appSlug: string;
  versionId: string;
};

export const AppVersionPageTabs = ['details', 'devices', 'test', 'activity'];

export type AppVersionPageQuery = AppVersionProps & {
  isPublic?: string;
  selectedTab?: typeof AppVersionPageTabs[number];
};

export type AppPageQuery = {
  appSlug: string;
};

export type AppVersionList = Array<{
  version: string;
  appVersions: AppVersion[];
}>;

export type AppVersion = {
  id: number;
  version: string;
  platform: string;
  buildNumber: number;
  buildSlug: string;
  lastUpdate: Date;
  description: string;
  whatsNew: string;
  minimumOs: string;
  minimumSdk: string;
  packageName: string;
  certificateExpires: Date;
  bundleId: string;
  size: number;
  supportedDeviceTypes: string[];
  distributionType: string;
  iconUrl: string;
  appName: string;
  promotionalText: string;
  keywords: string;
  reviewNotes: string;
  supportUrl: string;
  marketingUrl: string;
  scheme: string;
  configuration: string;
  publicInstallPageURL: string;
};

export interface PageContext extends NextContext {
  store: Store;
  isServer: boolean;
  req: Request;
}

export type Platform = 'ios' | 'android';
