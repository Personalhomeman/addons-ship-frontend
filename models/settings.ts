export const AppSettingsPageTabs = ['general', 'notifications'];

export type AppSettingsPageQuery = {
  appSlug: string;
  selectedTab?: typeof AppSettingsPageTabs[number];
};

export type IosSettings = {
  artifactExposingWorkflows: string;
  appleDeveloperAccountEmail: string;
  appSku: string;
  appSpecificPassword: string;
  selectedProvProfile: any;
  selectedCertificate: any;
};

export type AndroidSettings = {
  artifactExposingWorkflows: string;
  track: string;
  selectedKeystoreFile: any;
  selectedServiceAccountJsonFile: any;
};

export type ProvProfile = { name: string };
export type Certificate = { name: string };
export type KeystoreFile = { name: string };
export type ServiceAccountJsonFile = { name: string };

export type Settings = {
  iosSettings?: IosSettings;
  androidSettings?: AndroidSettings;
  provProfiles?: ProvProfile[];
  certificates?: Certificate[];
  keystoreFiles?: KeystoreFile[];
  serviceAccountJsonFiles?: ServiceAccountJsonFile[];
  projectType:
    | 'xamarin'
    | 'ios'
    | 'osx'
    | 'macos'
    | 'android'
    | 'cordova'
    | 'ionic'
    | 'react-native'
    | 'flutter'
    | 'other';
};

export type AppContact = {
  id: string;
  email: string;
  isConfirmed: boolean;
  notificationPreferences: {
    newVersion: boolean;
    successfulPublish: boolean;
    failedPublish: boolean;
  };
  isMarkedForDelete?: boolean;
  isMarkedForUpdate?: boolean;
};
