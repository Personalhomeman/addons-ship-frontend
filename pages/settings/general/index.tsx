import { Component } from 'react';
import { connect } from 'react-redux';

import { RootState } from '@/store';

type Props = {};

export class General extends Component<Props> {
  render() {
    return <div>general</div>;
  }
}

const mapStateToProps = ({ testDevices }: RootState) => ({ testDevices });

export default connect(mapStateToProps)(General as any);
