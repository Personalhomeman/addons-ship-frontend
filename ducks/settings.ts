import { Dispatch } from 'redux';
import { createAction, createReducer } from 'deox';

import { ShipAPIService } from '@/services/ship-api';
import { RootState } from '@/store';
import { actionTypeCreator } from '@/utils';
import { Settings } from '@/models/settings';

const $ = actionTypeCreator('SETTINGS');

const _fetchSettings = (appSlug: string) => async (
  dispatch: Dispatch,
  _getState: () => RootState,
  { shipApi }: { shipApi: ShipAPIService }
) => {
  dispatch(fetchSettings.next());

  try {
    const settings = await shipApi.getSettings(appSlug);

    dispatch(fetchSettings.complete(settings));
  } catch (error) {
    dispatch(fetchSettings.error(error));
  }
};

export const fetchSettings = Object.assign(_fetchSettings, {
  next: createAction($`GET_NEXT`),
  complete: createAction($`GET_COMPLETE`, resolve => (settings: Settings) => resolve(settings)),
  error: createAction($`GET_ERROR`, resolve => (error: Error) => resolve(error))
});

const _updateSettings = (settings: Settings) => async (
  dispatch: Dispatch,
  _getState: () => RootState,
  { shipApi }: { shipApi: ShipAPIService }
) => {
  dispatch(updateSettings.next());

  try {
    await shipApi.updateSettings(settings);

    dispatch(updateSettings.complete(settings));
  } catch (error) {
    dispatch(fetchSettings.error(error));
  }
};

export const updateSettings = Object.assign(_updateSettings, {
  next: createAction($`UPDATE_NEXT`),
  complete: createAction($`UPDATE_COMPLETE`, resolve => (settings: Settings) => resolve(settings)),
  error: createAction($`UPDATE_ERROR`, resolve => (error: Error) => resolve(error))
});

const defaultState: SettingsState = {
  projectType: 'other'
};

export type SettingsState = Settings;
export const settings = createReducer(defaultState, handleAction => [
  handleAction(fetchSettings.complete, (_, { payload }) => payload)
]);
