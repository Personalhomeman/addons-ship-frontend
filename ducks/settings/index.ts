import { createReducer } from 'deox';

import { Settings, AppContact } from '@/models/settings';

import fetchSettings from './fetchSettings';
import updateSettings from './updateSettings';
import addAppContact from './addAppContact';
import listAppContacts from './listAppContacts';
import updateAppContact from './updateAppContact';

const defaultState: SettingsState = {
  settings: {
    projectType: 'other'
  },
  appContacts: []
};

export { fetchSettings, updateSettings, addAppContact, listAppContacts, updateAppContact };

export type SettingsState = { settings: Settings; appContacts: AppContact[] };
export default createReducer(defaultState, handleAction => [
  handleAction(fetchSettings.complete, (state, { payload }) => ({ ...state, settings: payload })),
  handleAction(addAppContact.complete, (state, { payload }) => ({
    ...state,
    appContacts: state.appContacts.concat(payload)
  })),
  handleAction(listAppContacts.complete, (state, { payload }) => ({
    ...state,
    appContacts: payload
  })),
  handleAction(updateAppContact.complete, ({ appContacts, ...state }, { payload }) => ({
    ...state,
    appContacts: appContacts.map(appContact => {
      if (appContact.id === payload.id) {
        return payload;
      }

      return appContact;
    })
  }))
]);
