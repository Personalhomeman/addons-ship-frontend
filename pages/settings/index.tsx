import { Component } from 'react';
import Link from 'next/link';
import { Base, Tabs, Tab, Divider } from '@bitrise/bitkit';
import startCase from 'lodash/startCase';

import { PageContext } from '@/models';
import { AppSettingsPageQuery, AppSettingsPageTabs } from '@/models/settings';
import { fetchSettings } from '@/ducks/settings';

import General from './general';
import Notifications from './notifications';

interface SettingsPageProps extends AppSettingsPageQuery {
  pagePath: string;
}

export class SettingsPage extends Component<SettingsPageProps> {
  static defaultProps = {
    selectedTab: 'general'
  };

  static async getInitialProps({ query, store, req }: PageContext) {
    const { appSlug, selectedTab = AppSettingsPageTabs[0] } = (query as unknown) as AppSettingsPageQuery;

    switch (selectedTab) {
      case 'general':
        await store.dispatch(fetchSettings(appSlug) as any);
        break;
    }

    const pagePath = req.path.replace(new RegExp(`/${selectedTab}$`), '');

    return { appSlug, selectedTab, pagePath };
  }

  tabContent = () => {
    const { selectedTab } = this.props;

    switch (selectedTab) {
      case 'general':
        return <General />;
      case 'notifications':
        return <Notifications />;
      default:
        return <h1>{selectedTab}</h1>;
    }
  };

  render() {
    const { pagePath, selectedTab } = this.props;

    const tab = (key: string) => (
      <Tab active={selectedTab === key}>
        <Link href={`${pagePath}/${key}`}>
          <a>{startCase(key)}</a>
        </Link>
      </Tab>
    );

    return (
      <Base paddingVertical="x10">
        <Base maxWidth={960}>
          <Tabs gap="x10">
            {tab('general')}
            {tab('notifications')}
          </Tabs>
        </Base>
        <Divider color="gray-2" direction="horizontal" />
        <Base maxWidth={960}>{this.tabContent()}</Base>
      </Base>
    );
  }
}

export default SettingsPage;
