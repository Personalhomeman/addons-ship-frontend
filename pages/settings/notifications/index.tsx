import React from 'react';
import { connect } from 'react-redux';
import isEqual from 'lodash/isEqual';

import { AppContact } from '@/models/settings';
import { RootState } from '@/store';
import { addAppContact, updateAppContact, deleteAppContact } from '@/ducks/settings';

import View from './view';

type Props = {
  appSlug: string;
  appContacts: AppContact[];
  addAppContact: typeof addAppContact;
  updateAppContact: typeof updateAppContact;
  deleteAppContact: typeof deleteAppContact;
};

type State = {
  hasModifications: boolean;
  updatedAppContacts: AppContact[];
};

export class NotificationSettings extends React.Component<Props, State> {
  state: State = {
    hasModifications: false,
    updatedAppContacts: this.props.appContacts
  };

  componentDidUpdate({ appContacts: prevAppContacts }: Props) {
    const { appContacts } = this.props;

    if (prevAppContacts.length !== appContacts.length) {
      this.setState({ updatedAppContacts: appContacts, hasModifications: false });
    }
  }

  onAddEmail = (email: string) => {
    const { appSlug, addAppContact } = this.props;

    addAppContact(appSlug, email);
  };

  onNotificationPreferenceChanged = (email: string, key: string, value: boolean) => {
    const updatedAppContacts = this.state.updatedAppContacts.map(contact => {
      if (contact.email === email) {
        return {
          ...contact,
          isMarkedForUpdate: true,
          notificationPreferences: {
            ...contact.notificationPreferences,
            [key]: value
          }
        };
      }

      return contact;
    });

    const hasModifications = !isEqual(updatedAppContacts, this.props.appContacts);

    this.setState({ updatedAppContacts, hasModifications });
  };

  onDeleteContact = (email: string) => {
    const updatedAppContacts = this.state.updatedAppContacts.map(contact => {
      if (contact.email === email) {
        return {
          ...contact,
          isMarkedForDelete: true,
          isMarkedForUpdate: false
        };
      }

      return contact;
    });

    this.setState({ updatedAppContacts, hasModifications: true });
  };

  onSave = () => {
    const { appSlug, updateAppContact, deleteAppContact } = this.props;
    const { updatedAppContacts, hasModifications } = this.state;

    if (!hasModifications) return;

    updatedAppContacts
      .filter(({ isMarkedForUpdate, isMarkedForDelete }) => isMarkedForUpdate || isMarkedForDelete)
      .forEach(({ isMarkedForUpdate, isMarkedForDelete, ...appContact }) => {
        if (isMarkedForUpdate) {
          updateAppContact(appSlug, appContact);
        }

        if (isMarkedForDelete) {
          deleteAppContact(appSlug, appContact);
        }
      });

    this.setState({ hasModifications: false });
  };

  onCancel = () => {
    const { appContacts } = this.props;

    this.setState({ updatedAppContacts: appContacts, hasModifications: false });
  };

  render() {
    const { updatedAppContacts, hasModifications } = this.state;

    const props = {
      appContacts: updatedAppContacts,
      onAddEmail: this.onAddEmail,
      onNotificationPreferenceChanged: this.onNotificationPreferenceChanged,
      onDeleteContact: this.onDeleteContact,
      onSave: hasModifications ? this.onSave : undefined,
      onCancel: this.onCancel
    };

    return <View {...props} />;
  }
}

const mapStateToProps = ({ settings: { appContacts } }: RootState) => ({ appContacts }),
  mapDispatchToProps = { addAppContact, updateAppContact, deleteAppContact };

export default connect(
  mapStateToProps,
  mapDispatchToProps
  // @ts-ignore
)(NotificationSettings);
