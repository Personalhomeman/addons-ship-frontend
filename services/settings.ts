import { AppVersion } from '@/models';
import { IosSettings, AndroidSettings } from '@/models/settings';

export class SettingsService {
  constructor() {}

  isComplete = (
    appVersion: AppVersion,
    { iosSettings, androidSettings }: { iosSettings: IosSettings; androidSettings: AndroidSettings }
  ) => {
    switch (appVersion.platform) {
      case 'ios': {
        return ![
          'appleDeveloperAccountEmail',
          'appSku',
          'appSpecificPassword',
          'selectedAppStoreProvisioningProfile',
          'selectedCodeSigningIdentity'
        ].find(key => !iosSettings[key]);
      }
      case 'android': {
        return !['track', 'selectedKeystoreFile', 'selectedServiceAccount'].find(key => !androidSettings[key]);
      }
    }

    return true;
  };
}

export default new SettingsService();
