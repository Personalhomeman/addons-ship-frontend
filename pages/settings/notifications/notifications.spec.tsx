import { shallow, ShallowWrapper } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import { AppContact } from '@/models/settings';

import { NotificationSettings, Props } from '.';

describe('NotificationSettings', () => {
  const defaultProps: Props = {
    appSlug: 'some-app',
    hasLoaded: true,
    appContacts: [
      {
        id: 'bit-bot',
        email: 'bit.bot@bitrise.io',
        isConfirmed: true,
        notificationPreferences: {
          newVersion: true,
          successfulPublish: true,
          failedPublish: true
        }
      },
      {
        id: 'purr-req',
        email: 'purr.request@bitrise.io',
        isConfirmed: false,
        notificationPreferences: {
          newVersion: false,
          successfulPublish: false,
          failedPublish: true
        }
      }
    ],
    listAppContacts: jest.fn() as any,
    addAppContact: jest.fn() as any,
    updateAppContact: jest.fn() as any,
    deleteAppContact: jest.fn() as any
  };

  beforeEach(() => {
    const { listAppContacts, addAppContact, updateAppContact, deleteAppContact } = defaultProps;

    (([listAppContacts, addAppContact, updateAppContact, deleteAppContact] as any) as jest.Mock[]).forEach(mock =>
      mock.mockReset()
    );
  });

  it('renders correctly', () => {
    const tree = shallowToJson(shallow(<NotificationSettings {...defaultProps} />));

    expect(tree).toMatchSnapshot();
  });

  test('componentDidUpdate', () => {
    const { appContacts } = defaultProps;
    const wrapper = shallow(<NotificationSettings {...defaultProps} />);

    wrapper.setProps({ appContacts: [...appContacts, { email: ' doesnt@matt.er' }] });

    const { state } = wrapper.instance();

    expect(state).toMatchSnapshot();
  });

  it('sets the email', () => {
    const { addAppContact, appSlug } = defaultProps,
      email = 'whatever@email.com';
    const wrapper = shallow(<NotificationSettings {...defaultProps} />);

    (wrapper.instance() as NotificationSettings).onAddEmail(email);

    expect(addAppContact).toHaveBeenCalledWith(appSlug, email);
  });

  it('does not addAppContact with an empty email', () => {
    const { addAppContact } = defaultProps;
    const wrapper = shallow(<NotificationSettings {...defaultProps} />);

    (wrapper.instance() as NotificationSettings).onAddEmail('');

    expect(addAppContact).not.toHaveBeenCalled();
  });

  it('does not addAppContact if the email is already in the list', () => {
    const {
      appContacts: [{ email }],
      addAppContact
    } = defaultProps;
    const wrapper = shallow(<NotificationSettings {...defaultProps} />);

    (wrapper.instance() as NotificationSettings).onAddEmail(email);

    expect(wrapper.state('error')).not.toBeUndefined();
    expect(addAppContact).not.toHaveBeenCalled();
  });

  test('onNotificationPreferenceChanged', () => {
    const { appContacts } = defaultProps;

    const wrapper = shallow(<NotificationSettings {...defaultProps} />);

    (wrapper.instance() as NotificationSettings).onNotificationPreferenceChanged(
      appContacts[0].email,
      'successfulPublish',
      false
    );

    expect(wrapper.state('updatedAppContacts')).toMatchSnapshot();
    expect(wrapper.state('hasModifications')).toBe(true);
  });

  test('onDeleteContact', () => {
    const { appContacts } = defaultProps;

    const wrapper = shallow(<NotificationSettings {...defaultProps} />);

    (wrapper.instance() as NotificationSettings).onDeleteContact(appContacts[0].email);

    expect(wrapper.state('updatedAppContacts')).toMatchSnapshot();
    expect(wrapper.state('hasModifications')).toBe(true);
  });

  describe('onSave', () => {
    let updateAppContact: any,
      deleteAppContact: any,
      appContacts: AppContact[],
      appSlug: string,
      wrapper: ShallowWrapper;

    beforeEach(() => {
      ({ updateAppContact, deleteAppContact, appContacts, appSlug } = defaultProps);

      wrapper = shallow(<NotificationSettings {...defaultProps} />);
    });

    it('should not call updateAppContact, deleteAppContact without modifications', () => {
      (wrapper.instance() as NotificationSettings).onSave();

      expect(updateAppContact).not.toHaveBeenCalled();
      expect(deleteAppContact).not.toHaveBeenCalled();
    });

    it('should call updateAppContact', () => {
      const [appContact] = appContacts;
      (wrapper.instance() as NotificationSettings).onNotificationPreferenceChanged(
        appContact.email,
        'successfulPublish',
        false
      );

      (wrapper.instance() as NotificationSettings).onSave();

      expect(updateAppContact).toHaveBeenCalledWith(appSlug, {
        ...appContact,
        notificationPreferences: { ...appContact.notificationPreferences, successfulPublish: false }
      });
    });

    it('should call deleteAppContact', () => {
      const [appContact] = appContacts;

      (wrapper.instance() as NotificationSettings).onDeleteContact(appContact.email);
      (wrapper.instance() as NotificationSettings).onSave();

      expect(deleteAppContact).toHaveBeenCalledWith(appSlug, appContact);
    });
  });

  test('onCancel', () => {
    const { appContacts } = defaultProps;

    const wrapper = shallow(<NotificationSettings {...defaultProps} />);

    (wrapper.instance() as NotificationSettings).onNotificationPreferenceChanged(
      appContacts[0].email,
      'successfulPublish',
      false
    );

    expect(wrapper.state('updatedAppContacts')).toMatchSnapshot();
    expect(wrapper.state('hasModifications')).toBe(true);

    (wrapper.instance() as NotificationSettings).onCancel();

    expect(wrapper.state('updatedAppContacts')).toEqual(appContacts);
    expect(wrapper.state('hasModifications')).toBe(false);
  });
});
