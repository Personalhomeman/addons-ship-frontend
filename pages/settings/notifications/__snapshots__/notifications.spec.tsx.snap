// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationSettings componentDidUpdate 1`] = `
Object {
  "hasModifications": false,
  "updatedAppContacts": Array [
    Object {
      "email": "bit.bot@bitrise.io",
      "id": "bit-bot",
      "isConfirmed": true,
      "notificationPreferences": Object {
        "failedPublish": true,
        "newVersion": true,
        "successfulPublish": true,
      },
    },
    Object {
      "email": "purr.request@bitrise.io",
      "id": "purr-req",
      "isConfirmed": false,
      "notificationPreferences": Object {
        "failedPublish": true,
        "newVersion": false,
        "successfulPublish": false,
      },
    },
    Object {
      "email": " doesnt@matt.er",
    },
  ],
}
`;

exports[`NotificationSettings onCancel 1`] = `
Array [
  Object {
    "email": "bit.bot@bitrise.io",
    "id": "bit-bot",
    "isConfirmed": true,
    "isMarkedForUpdate": true,
    "notificationPreferences": Object {
      "failedPublish": true,
      "newVersion": true,
      "successfulPublish": false,
    },
  },
  Object {
    "email": "purr.request@bitrise.io",
    "id": "purr-req",
    "isConfirmed": false,
    "notificationPreferences": Object {
      "failedPublish": true,
      "newVersion": false,
      "successfulPublish": false,
    },
  },
]
`;

exports[`NotificationSettings onDeleteContact 1`] = `
Array [
  Object {
    "email": "bit.bot@bitrise.io",
    "id": "bit-bot",
    "isConfirmed": true,
    "isMarkedForDelete": true,
    "isMarkedForUpdate": false,
    "notificationPreferences": Object {
      "failedPublish": true,
      "newVersion": true,
      "successfulPublish": true,
    },
  },
  Object {
    "email": "purr.request@bitrise.io",
    "id": "purr-req",
    "isConfirmed": false,
    "notificationPreferences": Object {
      "failedPublish": true,
      "newVersion": false,
      "successfulPublish": false,
    },
  },
]
`;

exports[`NotificationSettings onNotificationPreferenceChanged 1`] = `
Array [
  Object {
    "email": "bit.bot@bitrise.io",
    "id": "bit-bot",
    "isConfirmed": true,
    "isMarkedForUpdate": true,
    "notificationPreferences": Object {
      "failedPublish": true,
      "newVersion": true,
      "successfulPublish": false,
    },
  },
  Object {
    "email": "purr.request@bitrise.io",
    "id": "purr-req",
    "isConfirmed": false,
    "notificationPreferences": Object {
      "failedPublish": true,
      "newVersion": false,
      "successfulPublish": false,
    },
  },
]
`;

exports[`NotificationSettings renders correctly 1`] = `
<_default
  appContacts={
    Array [
      Object {
        "email": "bit.bot@bitrise.io",
        "id": "bit-bot",
        "isConfirmed": true,
        "notificationPreferences": Object {
          "failedPublish": true,
          "newVersion": true,
          "successfulPublish": true,
        },
      },
      Object {
        "email": "purr.request@bitrise.io",
        "id": "purr-req",
        "isConfirmed": false,
        "notificationPreferences": Object {
          "failedPublish": true,
          "newVersion": false,
          "successfulPublish": false,
        },
      },
    ]
  }
  onAddEmail={[Function]}
  onCancel={[Function]}
  onDeleteContact={[Function]}
  onNotificationPreferenceChanged={[Function]}
/>
`;
